finalized api challenge (#1)
lewagon/data-certification-api: README.md

# Data certification API

Le Wagon Data Science certification exam starter pack for the predictive API test.

**💡 This challenge is completely independent of other challenges. It is not required to complete any other challenge in order to work on this challenge.**

## Setup

### Duplicate the repository for the API challenge

**📝 Let's duplicate the repository of the API challenge.**

Go to https://github.com/lewagon/data-certification-api:
- Click on `Use this template`
- Enter the repository name `data-certification-api`

> - Set as **Public**

Reviewer: ```suggestion
- Set it as **Public**
```

- Click on `Create repository from template`
- Click on `Code`
- Select `SSH`

> - Copy the SSH URL of the repository (the format is `git@github.com:YOUR_NAME/data-certification-api.git`)

Reviewer: ```suggestion
- Copy the SSH URL of the repository (the format is `git@github.com:YOUR_GITHUB_NICKNAME/data-certification-api.git`)
```

### Clone the repository for the API challenge

**📝 Now we will clone your new repository.**

Open your terminal and run the following commands:

👉 replace `YOUR_GITHUB_NICKNAME` with your **github nickname** and `PASTE_REPOSITORY_URL_HERE` with the SSH URL you just copied:

``` bash
cd ~/code/YOUR_GITHUB_NICKNAME
git clone PASTE_REPOSITORY_URL_HERE
cd data-certification-api
```

### Look around

**💡 The content of the challenge should look like this:**

``` bash
tree
```

```
.
├── Dockerfile
├── MANIFEST.in
├── Makefile
├── README.md
├── api
│   ├── __init__.py
│   └── app.py
├── exampack
│   ├── __init__.py
│   ├── data
│   ├── models
│   ├── predictor.py
│   ├── tests
│   │   └── __init__.py
│   └── utils.py
├── notebooks
├── requirements.txt
├── scripts
│   └── exampack-run
└── setup.py
```

> You can now open you favorite text editor and proceed with the challenges.

Reviewer: ```suggestion
Open your favourite text editor and proceed with the challenge.
```

## API challenge

**📝 In this challenge, you are provided with a trained model saved as `model.joblib`. The goal is to create an API that will predict the popularity of a song based on its other features.**

👉 In this challenge you will only need to edit the code of the API in `api/app.py` 🚨

> 👉 The package versions listed in `requirements.txt` should work out of the box will the pipeline saved in `model.joblib`

Reviewer: ```suggestion
👉 The package versions listed in `requirements.txt` should work out of the box with the pipelined model saved in `model.joblib`
```

### Install the required packages

**📝 Let's install the required packages.**

> 👉 The `requirements.txt` file lists the exact version of the packages required in order to be able to load the trained pipeline that we provide

Reviewer: ```suggestion
👉 The `requirements.txt` file lists the exact version of the packages required in order to be able to load the pipelined model that we provide
```

``` bash
pip install -r requirements.txt
```

If you encounter an version conflict while installing the packages, you will need to work in a new virtual environment:

👉 Only execute this commands if you encounter an issue while installing the packages 🚨

``` bash
pyenv install 3.8.6
pyenv virtualenv 3.8.6 certif
pyenv local certif
pip install -r requirements.txt

> ```

Reviewer: make it collapsable please

## API endpoint

> **📝 Let's begin by creating a prediction endpoint.**

Reviewer: to be removed

``` json
{
  "pred": 22.245404331801364
}
```

### Handle the API parameters

**📝 Add to your `/predict` endpoint the parameters that are required in order to make a prediction. Then create a `DataFrame` variable named `X_pred` from these parameters.**

👉 Remember that `FastAPI` only provides `string` variables as parameters to the endpoint

👉 Remember that the parameters need to be built into the `DataFrame` in the exact **same order** as the one required by the pipeline 🚨

💡 One way of building a `DataFrame` from individual values is to create it from a dictionary of lists

👉 Once `X_pred` is built, you may `print` it and have a look at the output of the `uvicorn` command in order to make sure that it is correctly created

### Make a prediction

**📝 Now that you have a trained pipeline and a `DataFrame` containing one observation, let's make a prediction and return it.**

👉 Remember that `FastAPI` can only handle simple data types as returned `json` values

Test your API using http://localhost:8000/docs and make sure that it works correctly before proceeding to put it in production.

> 

Reviewer: ### API specification

**Predict the popularity of a Spotify song**

`GET /Predict`

| Parameters | Description |
|---|---|
| acousticness | floating number for acousticness |
| danceability | todo |
| duration_ms | todo |
| energy | todo |
| explicit | todo |
| id | todo |
| instrumentalness | todo |
...

Returns a dictionary with the `artist`, the `name` of the song and predicted `popularity` as an integer.

Example request:

```
/predict?acousticness=0.654&danceability=0.499&duration_ms=219827&energy=0.19&explicit=0&id=0B6BeEUd6UwFlbsHMQKjob&instrumentalness=0.00409&key=7&liveness=0.0898&loudness=-16.435&mode=1&name=Back%20in%20the%20Goodle%20Days&release_date=1971&speechiness=0.0454&tempo=149.46&valence=0.43&artist=John%20Hartford
```

Example response:
```json
{
  "artist": "John Hartford",
  "name": "Back in the Goodle Days",
  "popularity": 28
}
  
```

## API in production

> **📝 Push your API to production on the hosting service of your choice.**

Reviewer: OK but the Makefile is for GCR only!!

Author: I added the instructions for GCP in a collapsed paragraph 🙏

👉 Remember to test your API locally and make an actual prediction before pushing your code to production in order to save time

👉 You may test your API with the following parameters (change the domain and port with your own):

```

> http://localhost:8000/predict?acousticness=0.654&danceability=0.499&duration_ms=219827&energy=0.19&explicit=0&id=0B6BeEUd6UwFlbsHMQKjob&instrumentalness=0.00409&key=7&liveness=0.0898&loudness=-16.435&mode=1&name=Back%20in%20the%20Goodle%20Days&release_date=1971&speechiness=0.0454&tempo=149.46&valence=0.43&artist=John%20Hartford

Reviewer: To be removed

```

💡 With a configured GCP account, the directives in the `Makefile` should be enough to push your API to production

> 👉 Remember to update the `GCP_PROJECT_ID` in the `Makefile` 🚨

Reviewer: Once you had change your `GCP_PROJECT_ID` in the `Makefile`, run the following command to build and deploy your containerized API to GC Container Registry and finally GCRun.

Author: I added the instructions to test locally then push and deploy 🙏
